src/metabase/query_processor/api.clj

```diff
@@ -62,7 +63,9 @@
       (when-not database
         (throw (ex-info (tru "`database` is required for all queries whose type is not `internal`.")
                         {:status-code 400, :query query})))
-      (api/read-check :model/Database database))
+      (when-not (or (mi/can-read? :model/Database database)
+                    (perms/user-has-published-table-permission-for-database? database))
+        (api/throw-403)))
```

Comment:
Should this logic just be in the `can-read?` method for `:model/Database`? That way someone else can't forget to do this.

`can-read?` for a Database basically just means they can see the `:model/Database` row itself so if that's true whenever they have published table perms then this logic should go there I think

Reply:
@camsaul we don't want the database to be user-visible in this case; `/api/database` should not return it if the user doesn't have normal data access to any of the tables. So we had to decide whether
1. `can-read?` is true for the published table case, and we need to filter out such databases in some cases
2. leave `can-read?` as is and make it more restrictive, and only elevate access when needed exactly.

For now, we decided to use approach 2.

Reply:
We considered this, and didn't want to bundle them.

Reply:
actually, maybe it's a sign that `can-read?` is/was misused? and instead it should have checked for `:perms/create-queries` >= `query-builder` more directly